lib/azure/armrest/armrest_manager.rb

```diff
@@ -209,14 +209,71 @@ def providers
 
       # Returns information about the specific provider +namespace+.
       #
-      def provider_info(namespace)
-        url = url_with_api_version(@base_url, 'providers', namespace)
+      def provider_info(provider)
+        url = url_with_api_version(@base_url, 'providers', provider)
         response = rest_get(url)
         JSON.parse(response.body)
       end
 
       alias geo_locations provider_info
 
+      # Returns a list of all locations for all resource types of the given
+      # +provider+. If you do not specify a provider, then the locations for
+      # all providers will be returned.
+      #
+      # If you need individual details on a per-provider basis, use the
+      # provider_info method instead.
+      #--
+      # TODO: I think a 7-day cache may be wise here as the overall list of
+      # locations is unlikely to change from day to day.
+      #
+      def locations(provider = nil)
+        array = []
+
+        if provider
+          info = provider_info(provider)
+          info['resourceTypes'].map{ |rt| array << rt['locations'] }
+        else
+          threads = []
+
+          providers.each do |provider_hash|
+            provider = provider_hash['namespace']
+            threads << Thread.new{
+              info = provider_info(provider)
+              info['resourceTypes'].map{ |rt| array << rt['locations'] }
+            }
+          end
+
+          threads.each(&:join)
+        end
+
+        array.flatten!
+        array.uniq!
+        array.delete("") # Seems to pickup blank elements for some reason.
+
+        array
+      end
+
+      # Returns an array of publishers for the given +region+ and +provider+,
+      # or "Microsoft.Compute" if no provider is given.
+      #
+      # Examples:
+      #
+      #   arm.publishers('eastus')
+      #   arm.publishers('eastus', 'Microsoft.ClassicCompute')
+      #
+      def publishers(region, provider = 'Microsoft.Compute', subscription_id = @subscription_id)
+        @api_version = '2015-06-15'
```

Comment:
No, the `@api_version` magic string :smile: 

In fact, it seems like it would be better to be able to change that without installing a new version of this gem.  But, that *might* have other implications.  Such as, if the `api_version` changes, then that might mean that the implementation of the gem would have to change to accommodate it.  Not sure, but possible?

Nonetheless, I'd like to see that not hidden in a method call inside the API and rather see it exposed somewhere obvious.  Maybe even in the gemspec ... that way, it's obvious from the gemspec what version of the ARM API is supported by this gem.